kore/src/Kore/Strategies/Verification.hs

```diff
@@ -264,54 +268,49 @@ verifyClaim
             strategy
             & Foldable.toList
             & Limit.takeWithin depthLimit
-    handle
-        handleLimitExceeded
-        $ Strategy.leavesM
+    proofDepths <-
+        Strategy.leavesM
             updateQueue
             (Strategy.unfoldTransition transit)
-            (limitedStrategy, startGoal)
+            (limitedStrategy, (ProofDepth 0, startGoal))
             & fmap discardStrategy
             & throwUnproven
+            & handle handleLimitExceeded
+    let maxProofDepth = sconcat (ProofDepth 0 :| proofDepths)
+    infoProvenDepth maxProofDepth
   where
     discardStrategy = snd
 
     handleLimitExceeded
-        :: Strategy.LimitExceeded CommonProofState
-        -> ExceptT (OrPattern VariableName) simplifier ()
-    handleLimitExceeded (Strategy.LimitExceeded patterns) =
-        Monad.Except.throwError
-        . OrPattern.fromPatterns
-        $ fmap
-            (ProofState.proofState lhsProofStateTransformer)
-            patterns
+        :: Strategy.LimitExceeded (ProofDepth, CommonProofState)
+        -> Verifier simplifier a
+    handleLimitExceeded (Strategy.LimitExceeded states) =
+        (Monad.Except.throwError . OrPattern.fromPatterns)
+        (ProofState.proofState lhsProofStateTransformer . snd <$> states)
 
     updateQueue = \as ->
         Strategy.unfoldSearchOrder searchOrder as
         >=> lift . Strategy.applyBreadthLimit breadthLimit discardStrategy
 
     throwUnproven
-        :: LogicT (Verifier simplifier) CommonProofState
-        -> Verifier simplifier ()
+        :: LogicT (Verifier simplifier) (ProofDepth, CommonProofState)
+        -> Verifier simplifier [ProofDepth]
     throwUnproven acts =
-        Logic.runLogicT acts throwUnprovenOrElse done
-      where
-        done = return ()
-
-    throwUnprovenOrElse
-        :: CommonProofState
-        -> Verifier simplifier ()
-        -> Verifier simplifier ()
-    throwUnprovenOrElse proofState acts = do
-        ProofState.extractUnproven proofState
-            & Foldable.traverse_
-                ( Monad.Except.throwError
-                . OrPattern.fromPattern
-                . getConfiguration
-                )
-        acts
+        do
+            (proofDepth, proofState) <- acts
+            let maybeUnproven = ProofState.extractUnproven proofState
+            Foldable.for_ maybeUnproven $ \unproven -> do
+                infoUnprovenDepth proofDepth
+                Monad.Except.throwError . OrPattern.fromPattern
+                    $ getConfiguration unproven
+            pure proofDepth
+        & Logic.observeAllT
```

Comment:
Would it be a good idea for me to use `runLogicT` here and get only the maximum value?
Something like
`Logic.runLogicT logic (fmap . (<>)) (pure $ ProofDepth 0)`
where `logic` is the `do` block

Reply:
It might be a good idea to do something else, but that will build a big thunk on the heap before it starts computing the maximum depth.

Reply:
On second thought, this doesn't seem to cause any change in performance, so I'd like to keep it simple until we see any problems.